scripts/migration/lib.sh

```diff
@@ -0,0 +1,286 @@
+CONFIG_FILES=(
+  "defaults/common-config.yaml"
+  "common-config.yaml"
+  "defaults/sc-config.yaml"
+  "sc-config.yaml"
+  "defaults/wc-config.yaml"
+  "wc-config.yaml"
+  "secrets.yaml"
+  ".state/kube_config_sc.yaml"
+  ".state/kube_config_wc.yaml"
+)
+
+declare -a SC_CONFIG_FILES
+SC_CONFIG_FILES=(
+  "defaults/common-config.yaml"
+  "defaults/sc-config.yaml"
+  "common-config.yaml"
+  "sc-config.yaml"
+)
+
+declare -a WC_CONFIG_FILES
+WC_CONFIG_FILES=(
+  "defaults/common-config.yaml"
+  "defaults/wc-config.yaml"
+  "common-config.yaml"
+  "wc-config.yaml"
+)
+
+# --- logging functions ---
+
+log_info_no_newline() {
+  echo -e -n "[\e[34mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_info() {
+  log_info_no_newline "${*}\n"
+}
+
+log_warn_no_newline() {
+  echo -e -n "[\e[33mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_warn() {
+  log_warn_no_newline "${*}\n"
+}
+
+log_error_no_newline() {
+  echo -e -n "[\e[31mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_error() {
+  log_error_no_newline "${*}\n"
+}
+
+log_fatal() {
+  log_error "${*}"
+  exit 1
+}
+
+# --- git version
+
+git_version() {
+  git -C "${ROOT}" describe --exact-match --tags 2> /dev/null || git -C "${ROOT}" rev-parse HEAD
+}
+
+# --- config functions ---
+
+# Usage: config_version <sc|wc>
+config_version() {
+  if [[ ! "${1:-}" =~ ^(sc|wc)$ ]]; then
+    log_fatal "usage: config_version <sc|wc>"
+  fi
+
+  local prefix="${1}"
+
+  local version
+  version="$(yq4 ".global.ck8sVersion" <<< "${CONFIG["${prefix}"]}")"
+
+  VERSION["${prefix}-config"]="${version}"
+  version="${version#v}"
+  VERSION["${prefix}-config-major"]="${version%%.*}"
+  version="${version#*.}"
+  VERSION["${prefix}-config-minor"]="${version%%.*}"
+  version="${version#*.}"
+  VERSION["${prefix}-config-patch"]="${version}"
+}
+
+# Usage: config_validation <secrets|sc|wc>
+config_validate() {
+  local defaults
+  local setmes
+  local pass="true"
+
+  IFS=$'\n'
+
+  case "${1:-}" in
+  secrets)
+    log_info "validating ${1}"
+
+    setmes="$(yq_paths "set-me" < <(sops -d "${CK8S_CONFIG_PATH}/secrets.yaml"))"$'\n'"$(yq_paths "somelongsecret" < <(sops -d "${CK8S_CONFIG_PATH}/secrets.yaml"))"
+    for setme in ${setmes}; do
+      log_error "error: \"${setme//\"/}\" is unset in ${1}"
+      pass="false"
+    done
+    ;;
+
+  sc|wc)
+    log_info "validating ${1}-config"
+
+    defaults="$(yq_merge "${CK8S_CONFIG_PATH}/defaults/common-config.yaml" "${CK8S_CONFIG_PATH}/defaults/${1}-config.yaml")"
+    setmes="$(yq_paths "set-me" <<< "${defaults}")"
+
+    for setme in ${setmes}; do
+      compare=$(diff <(yq4 -oj "${setme}" <<< "${defaults}") <(yq4 -oj "${setme}" <<< "${CONFIG["${1}"]}") || true)
+      if [[ -z "${compare}" ]]; then
+          log_error "error: \"${setme//\"/}\" is unset in ${1}-config"
+          pass="false"
+      fi
+    done
+    ;;
+
+  *)
+    log_fatal "usage: config_validate <secrets|sc|wc>"
+    ;;
+  esac
+
+  if [[ "${pass}" = "false" ]]; then
+    exit 1
```

Comment:
Do we always want to do a hard stop here or do we want to prompt (as the bin variant does)?

Reply:
I can add some optional prompt, but we have to be careful to disable that in automation.

Reply:
Think this should suffice, though I don't know if it will trigger in automation tools. It shouldn't hopefully.